src/compiler/generate-common-tests.xsl

```diff
@@ -266,8 +266,8 @@
          <xsl:with-param name="local-name" select="generate-id()"/>
          <xsl:with-param name="last" select="empty(following-sibling::x:scenario)"/>
          <xsl:with-param name="params" as="element(param)*">
-            <xsl:for-each select="$vars">
-               <param name="{ @name }" select="${ @name }"/>
+            <xsl:for-each select="distinct-values($vars/@name)">
```

Comment:
Strictly speaking, this has to be able to eliminate different manifestation of the same name (taking different prefix and also EQName into account). But in practice, I agree to the current implementation.

Reply:
You're right about the different manifestations. I'd like to fix this, but I won't have time today. If by some chance you are ready to merge this PR before I fix it, then I'll log it as a bug and fix it separately.

Reply:
It'll take some more days for me to review this pr.
I think the current implementation is good enough for the vast majority of use cases. We can just log an issue or comment the code about the limitation, and even leave it open until someone raises a fix request.

Reply:
Another thing I'd like to add is a clear error message if someone tries to use the reserved "result" variable name for a user-defined XSpec variable. Currently, what I'm seeing is a nonideal error message for a local variable named result in the XSpec namespace, while a global variable by that name would be shadowed silently.

Reply:
Commit 5115b07d19a8f4c4a8c94b0e6c5a04391397900d adds error checking for a user-defined XSpec variable named "result" in the XSpec namespace. There may be some subtlety I've missed, so I'll appreciate any skepticism you bring to the review.

Error cases I ran manually, to confirm that the error message appears in each case:
- Added `<x:variable name="x:result" select="'abc'"/>` to a test as child of either `description` or `scenario` (pending or not).
- Same as above, but using a different prefix for the XSpec namespace.
- Added `<x:variable name="Q{http://www.jenitennison.com/xslt/xspec}result" select="'abc'"/>` to a test as child of either `description` or `scenario` (pending or not).
- Added `<variable name="result" select="'abc'"/>` in `xspec-no-prefix.xspec` as child of either `description` or `scenario` (pending or not). I didn't think the name was interpreted as being in the default namespace, but evaluating `namespace-uri-from-QName(resolve-QName(//*:variable/@name,//*:variable))` against that file returns `http://www.jenitennison.com/xslt/xspec`.

Reply:
Do we need to restrict the name check only to `x:result`? I think not only `x:result` but also any `http://www.jenitennison.com/xslt/xspec` name in general should be considered as reserved.

Reply:
> I didn't think the name was interpreted as being in the default namespace

The name should be resolved like these? https://github.com/xspec/xspec/blob/d48809471140122abf3154ff82e1ca52054a8276/src/compiler/generate-query-helper.xsl#L28-L31
https://github.com/xspec/xspec/blob/d48809471140122abf3154ff82e1ca52054a8276/src/compiler/generate-tests-helper.xsl#L31-L33

If so, we may want to create a common function in [`xspec-utils.xsl`](https://github.com/xspec/xspec/blob/master/src/common/xspec-utils.xsl) and reuse it.

And maybe `@preserve-space` has a similar issue:
https://github.com/xspec/xspec/blob/d48809471140122abf3154ff82e1ca52054a8276/src/compiler/generate-common-tests.xsl#L140-L141

Reply:
I uploaded 3b76092e1a5830cb05da402e62d9768d353406cf, which broadens the detection of reserved names to anything in the XSpec namespace and tries to correct the way of gathering distinct names of variables.

When I considered namespace usage that the code might need to support (e.g., if the test author binds a particular prefix to one URI in one scenario and a different URI in a different scenario, do the compiled variable definitions and references come out right?), I found that some parts of the compiler code seem to assume that all the necessary namespace declarations occur on the `x:description` element. For example, overriding a global parameter using `<x:param name="e:globalparam" xmlns:e="...">` does not work if `x:description` does not have xmlns:e="...". Is that supposed to work? What about prefix-qualified names of functions, templates, or modes in the system being tested? Should I log an issue?

So far, I've made my code adopt the same limitation that I think I'm seeing elsewhere, by assuming that namespace prefixes in any `x:variable/@name` attribute values have the associated namespace declarations on the `x:description` element. I don't think I will have time to remove this limitation in the next week. Maybe the limitation can be removed for XSpec variables at the same time that the analogous limitation is removed for other areas of XSpec. If you have a better suggestion or if you think my observations are wrong, let me know.

I worked on this before I saw your last comment above, and I'll look at those file references shortly.

Reply:
> I found that some parts of the compiler code seem to assume that all the necessary namespace declarations occur on the `x:description` element.

Right. The current implementation assumes so. I presume it to be a bug rather than an elaborate design. Somehow I haven't seen a complaint about the behavior. So I agree to the same limitation in `x:variable`. AFAIK, `x:import` has a similar problem.

If you like to mention the limitation in your code, you can file an issue and refer to the issue number in a code comment.

Reply:
> > I didn't think the name was interpreted as being in the default namespace
> 
> The name should be resolved like these?

No, my case is different because variables should support URI-qualified names.

(Actually, I don't know why the `named-templates` key shouldn't support URI-qualified names of templates. I can't find any reference to that key, though. Maybe it is unneeded code.)

I believe the treatment of `preserve-space` that you highlighted is OK because the data type of that attribute in the schema is `xsd:NMTOKENS`, which precludes a value like `Q{}pre` or `Q{...}pre`.

What I really meant when I wrote "I didn't think the name was interpreted as..." was: I didn't think the name was _supposed to be_ interpreted as being in the default namespace, because I didn't think the default namespace applied to an attribute value.

Reply:
In other words, I thought https://github.com/galtm/xspec/blob/variablesupport/src/compiler/generate-common-tests.xsl#L660 might have to be `<xsl:when test="namespace-uri-from-QName(if (contains(@name, ':')) 
 	            then resolve-QName(@name, .) 
 	            else QName('', @name)) eq $xspec-namespace">`.